### ISSUE TITLE
Incremental Scala Compilation can fail when moving a class in a separate module

### ISSUE THREAD

**Issue by KrauseStefan** [gradle/gradle#8421]
Incremental Scala Compilation can fail when moving a class in a separate module

I created a simplyfied setup with with two scala modules (**A** and **B**), where module **A** depends on module **B** with a compileOnly dependency.

Module **B** has two classes a Service class and a DTO class.
Module **A** has a single class that only imports the service not the DTO from module **B**

Module **A** does use DTO, but the type is not applied directly since it is used in a `foreach` construct

``` scala
import module.b.services.Service

object ModuleA extends App {
  new Service().getDto().foreach(dto => println(dto.message))
}
```

The service in Module **B** is defined as
``` scala
// import module.b.moved.dtos.MyDTO
import module.b.dtos.MyDTO

class Service {
  def getDto(): Option[MyDTO] = Some(new MyDTO)
}
```

The problem happens when the DTO is moved, if you see the commented out line above.
Gradle correctly rebuilds Module **B** but module **A** is not rebuild and running it fails with the following error:
``` 
Exception in thread "main" java.lang.NoClassDefFoundError: module/b/dtos/MyDTO
        at com.module.a.ModuleA$$anonfun$1.apply(ModuleA.scala:6)
        at scala.Option.foreach(Option.scala:257)
        at com.module.a.ModuleA$.delayedEndpoint$com$module$a$ModuleA$1(ModuleA.scala:6)
        at com.module.a.ModuleA$delayedInit$body.apply(ModuleA.scala:5)
        at scala.Function0$class.apply$mcV$sp(Function0.scala:40)
        at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:12)
        at scala.App$$anonfun$main$1.apply(App.scala:76)
        at scala.App$$anonfun$main$1.apply(App.scala:76)
        at scala.collection.immutable.List.foreach(List.scala:381)
        at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:35)
        at scala.App$class.main(App.scala:76)
        at com.module.a.ModuleA$.main(ModuleA.scala:5)
        at com.module.a.ModuleA.main(ModuleA.scala)
Caused by: java.lang.ClassNotFoundException: module.b.dtos.MyDTO
        at java.net.URLClassLoader.findClass(Unknown Source)
        at java.lang.ClassLoader.loadClass(Unknown Source)
        at sun.misc.Launcher$AppClassLoader.loadClass(Unknown Source)
        at java.lang.ClassLoader.loadClass(Unknown Source)
        ... 13 more
```

Checking dependencies with jdeps I get the following output for one class file
``` 
jdeps.exe '.\module-a\build\classes\scala\main\com\module\a\ModuleA$$anonfun$1.class'
ModuleA$$anonfun$1.class -> not found
ModuleA$$anonfun$1.class -> C:\Program Files\Java\jdk1.8.0_171\jre\lib\rt.jar
   com.module.a (ModuleA$$anonfun$1.class)
      -> java.lang
      -> module.b.dtos                               not found
      -> scala                                       not found
      -> scala.runtime                               not found
```
       
The problem was originally experienced in a big JAVA EE application on Wildfly.
My reproduction can be found here: https://github.com/KrauseStefan/gradle_5.2_issue_scala_incremental

I have two branches in this repository, switching between the branches moved the DTO class.
The below commands are all run from the root of the repository in Powershell.

Running building and running fails if a clean build succeed on the other branch
`.\gradlew.bat build; .\run.cmd`

A clean build always succeeds 
`.\gradlew.bat clean; .\gradlew.bat build; .\run.cmd`

### Expected Behavior

I expect incremental builds to work also when moving DTOs that the src does not directly depend on.

### Your Environment

build scan URLs:
 * Clean build master branch: https://gradle.com/s/g6j33p3qd7usg 
 * Incremental Build dependencyMoved branch: https://gradle.com/s/ua3ogagqtspju

**Comment by KrauseStefan** [comment]
I see I put in an older version of the Scala framework, I just tried it again with Scala 2.12.8 samme issue

**Comment by KrauseStefan** [comment]
I guess the problem the problem is that no change is made in A that would trigger a compilation, Gradle seems to relay on that the import would be updated when a package is moved.

The Scala plugin in Gradle would somehow need to trigger recompilation of the module if any implicit dependency has been moved, I imagine that this kind of issue would also occur if the dependency was a maven dependency.

I imagine that doing this would require that Gradle would need to test if a package that previous build ended up in depending on has moved... I guess the way to do that would be to add the moved file in module B to the input list of module A during an incremental build, then any change to that file would trigger a new build as well as local files.
I realize that this introduces some problems as well, now the will be a difference between the input of a clean build and an incremental build, I don't know a good way of solving that, besides compiling twice or have some uncacheable preanalysis step. 

I do not know an if a integration with the compiler through **Bloop** could help solve this.
I however fear that the issue lies in an incompatibility between the way Gradle does incremental builds and the Scala language.

Do you have any ideas @big-guy?

**Comment by big-guy** [comment]
Thanks for the reproduction case.  I suspect this is something happening in Zinc as Gradle delegates to Zinc for incremental compilation.  We pass everything to the Zinc compiler and it keeps a separate analysis file it uses for determining what's changed.  It makes sense that clean builds fix it because that deletes the analysis file. 

The version of Zinc we integrate with is pretty old, so we're trying to merge https://github.com/gradle/gradle/pull/8485 for 5.4. I'm not sure if that will fix this problem.

**Comment by KrauseStefan** [comment]
Let me know if there is anything you need me to do.

**Comment by melix** [comment]
I could confirm that the bug is fixed by #8485.
